crates/analyzer-core/src/lsp_position_struct.rs

```diff
@@ -0,0 +1,221 @@
+use analyzer_abstractions::lsp_types::{self, Position, TextDocumentContentChangeEvent};
+
+#[derive(Clone, Debug)]
+pub struct LspPos {
+	eof: bool, // flag indicates if the last character in file is '\n'
+	ranges: Vec<usize>,
+}
+
+impl LspPos {
+	// helper function
+	fn parse_string(string: &String) -> (Vec<usize>, bool) {
+		let mut result: Vec<usize> = Vec::new();
+		if string.is_empty() {
+			return (result, false);
+		}
+
+		let bytes = string.as_bytes();
+		for i in 0..bytes.len() {
+			if bytes[i] == b'\n' {
+				result.push(i);
+			}
+		}
+		// If there are bytes left add it to vector
+		if *result.last().unwrap_or(&(usize::MAX - 1)) != bytes.len() - 1 {
+			result.push(bytes.len() - 1);
+		}
+
+		let eof = string.as_bytes().last() == Some(&b'\n');
+
+		(result, eof)
+	}
+
+	pub fn parse_file(file: &String) -> Self {
+		let parse = LspPos::parse_string(&file);
+		LspPos { ranges: parse.0, eof: parse.1 }
+	}
```

Comment:
You can make use of the tuple type returned from `parse_string(...)` here to make the code more concise:

```rust
let (ranges, eof) = LspPos::parse_string(&file);
LspPos { ranges, eof }
```